File: src/2608.py
from collections import deque

class Solution:
    def findShortestCycle(self, n: int, edges: list[list[int]]) -> int:
        graph = [[] for _ in range(n)]
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)

        INF = 10 ** 9
        answer = INF

        # перебираем каждую вершину как стартовую для bfs
        for start in range(n):
            # треугольник - минимальный цикл
            if answer == 3:
                return 3

            dist = [-1] * n
            parent = [-1] * n
            queue = deque([start])
            dist[start] = 0

            while queue:
                u = queue.popleft()
                for v in graph[u]:
                    if dist[v] == -1:  # ещё не посещён
                        dist[v] = dist[u] + 1
                        parent[v] = u
                        queue.append(v)
                    elif v != parent[u]:  # сосед не родитель -> это новый путь
                        # до уже посещённой вершины -> нашли цикл
                        cycle_len = dist[u] + dist[v] + 1
                        if cycle_len < answer:
                            answer = cycle_len

        return answer if answer != INF else -1

File: src/2608.kt
class Solution {
    fun findShortestCycle(n: Int, edges: Array<IntArray>): Int {
        val adj = Array(n) { mutableListOf<Int>() }
        for ((u, v) in edges) {
            adj[u].add(v)
            adj[v].add(u)
        }

        var minCycle = Int.MAX_VALUE

        for (start in 0 until n) {
            val dist = IntArray(n) { -1 }
            val parent = IntArray(n) { -1 }
            val queue = ArrayDeque<Int>()

            dist[start] = 0
            queue.add(start)

            while (queue.isNotEmpty()) {
                val u = queue.removeFirst()

                if (dist[u] * 2 + 1 >= minCycle) break

                for (v in adj[u]) {
                    if (dist[v] == -1) {
                        dist[v] = dist[u] + 1
                        parent[v] = u
                        queue.add(v)
                    } else if (v != parent[u]) {
                        minCycle = minOf(minCycle, dist[u] + dist[v] + 1)
                    }
                }
            }
        }
        return if (minCycle == Int.MAX_VALUE) -1 else minCycle
    }
}